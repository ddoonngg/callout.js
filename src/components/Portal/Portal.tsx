import * as React from 'react';
import ReactDOM from 'react-dom';

type Container = HTMLElement | (() => HTMLElement);

function getContainer(contianer: Container): HTMLElement {
  if (typeof contianer === 'function') {
    return contianer();
  }
  return contianer;
}

interface PortalProps {
  children: React.ReactNode;
  container?: Container;
  disabledPortal?: boolean;
}

const Portal: React.FunctionComponent<PortalProps> = (props) => {
  const { container, children, disabledPortal = false } = props;
  const [mountNode, setMountNode] = React.useState(null);

  React.useEffect(() => {
    if (!disabledPortal) {
      setMountNode(getContainer(container) || document.body);
    }
  }, [disabledPortal, container]);

  if (disabledPortal) {
    if (React.isValidElement(children)) {
      return React.cloneElement(children);
    }
  }
  return mountNode ? ReactDOM.createPortal(children, mountNode) : mountNode;
};

Portal.displayName = 'Portal';

export default Portal;
