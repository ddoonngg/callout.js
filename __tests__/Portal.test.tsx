/* eslint-disable max-nested-callbacks */
import * as React from 'react';
import ReactDOM from 'react-dom';
import Portal from '../src/components/Portal';
import { render, fireEvent, waitFor, screen } from '@testing-library/react';

describe('<Portal/>', () => {
  describe('container', () => {
    it(`should render to the parent element if 'disabledPortal' is true`, () => {
      const { container } = render(
        <Portal disabledPortal>
          <h1>children</h1>
        </Portal>
      );
      expect(container.firstChild).toMatchSnapshot(`<h1>children</h1>`);
    });
  });
});
