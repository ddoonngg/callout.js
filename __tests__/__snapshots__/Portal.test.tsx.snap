// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Portal/> container should render to the parent element if 'disabledPortal is true: <h1>children</h1> 1`] = `
<h1>
  children
</h1>
`;

exports[`<Portal/> container should render to the parent element if 'disabledPortal' is true: <h1>children</h1> 1`] = `
<h1>
  children
</h1>
`;
